### ISSUE TITLE
Allow to reopen pull requests after a force push

### ISSUE THREAD

**Issue by cirosantilli** [isaacs/github#361]
If I close a pull request, force push to the branch, and then try to reopen it, the "Reopen pull request button" is deactivated and the on hover tooltip says "The XXX branch was force pushed or recreated".

Why does it behave like this, considering that you can force push open pull requests and it all works as expected?

Sample: https://github.com/plasticboy/vim-markdown/pull/176

**Comment by stuartpb** [comment]
Yeah, that's weird. I can't think of any reason why that should happen. Maybe it's supposed to force people to keep from re-using pull requests for multiple merges (like a pull request can only be merged once)?

Did you email GitHub about this? What was their response?

**Comment by cirosantilli** [comment]
@stuartpb I've emailed them as usual, but no reply so far. When they do reply I'll update.

**Comment by cirosantilli** [comment]
github replied

**Comment by jayphelps** [comment]
:+1: I don't understand how this is different either. I've literally done hundreds (if not thousands) of force pushes to PRs, without issue, but if it's closed a force push seals it shut forever no matter what.

**Comment by TheSeldonPlan** [comment]
+1 on this issue, any resolution?

**Comment by nhooey** [comment]
@cirosantilli What did GitHub say in their reply?

**Comment by stuartpb** [comment]
From the OP:

> We're blocking the pull request reopen if the current head isn't a descendant of the stored head sha (which is what the head was when the pull request was closed). We are not allowing the reopen in that case, because there is no good way to tell what changes have happened while a pull request was closed and the head branch has changed.

**Comment by stuartpb** [comment]
I think their point is that Github tracks every HEAD change while the PR is open (for "commented on an outdated diff" purposes), but doesn't do this tracking when the pull request is closed, so they can't construct a history between what the branch's HEAD was when the PR was last open and what the HEAD of the PR's branch is now.

I agree with @cirosantilli that this is ridiculous, since they don't use this kind of history tracking anywhere (at least not in the public-facing UI) except for "outdated diff" comments, which are already associated to a specific commit and point in time.

**Comment by stuartpb** [comment]
And, if I understand their complaint correctly, it should be possible to work around this by re-pushing the PR's original HEAD (the HEAD it had when the PR was closed), re-opening the PR, then re-force-pushing the HEAD it had before pushing the HEAD from when it was closed. (Which is *completely stupid* to have to do.)

**Comment by alexmuller** [comment]
This would be useful to have (and I'm sure it used to work).

I think the use case that @cirosantilli has is quite common. After receiving feedback on a pull request it's nice to be able to close it to indicate that it's not ready to be merged, and then modify the commits and force push to keep the history tidy.

**Comment by paulp** [comment]
The practice of updating PRs in place with force pushes should be [reconsidered](https://github.com/suffuse/go-suffuse/issues/30) anyway. It was inevitable that practices which amount to telling lies for a more palatable history would carry consequences (just like in real life.)

**Comment by stuartpb** [comment]
[It's not just "telling lies for a more palatable history".](http://blog.izs.me/post/37650663670/git-rebase)

**Comment by cirosantilli** [comment]
let's not flamewar to rebase or not to rebase.

It's all just electrons moving around inside our CPUs in the end.

:smile:

**Comment by mtbc** [comment]
Yes, force-pushing the original HEAD doesn't seem to help, the PR still can't be reopened. From our own fork GH should be assuming that we know what we're doing.

**Comment by shailendrabirthare** [comment]
I was in similar situation today i.e. used "Merge Pull Request" button which by default merge the changes into target branch and close the PR. I later discovered a bug in testing which I wanted the original developer to fix. I wanted a way to reopen this PR so that more commits can be added to this same PR but could not as there is no button for reopening the PR.

**Comment by jakirkham** [comment]
GitHub reply:

> We're blocking the pull request reopen if the current head isn't a descendant of the stored head sha (which is what the head was when the pull request was closed). We are not allowing the reopen in that case, because there is no good way to tell what changes have happened while a pull request was closed and the head branch has changed.

However, they don't seem to allow this even if you force push and return to the same sha. That seems like it should be allowed, no?

**Comment by jakirkham** [comment]
Workaround. Pushed the branch (oops its not on the same commit). Steps to get PR reopened.

1. Delete the branch using the button in the GitHub PR.
2. Move the branch back to where the last PR sha is locally.
3. Push the branch again.
4. Reopen the PR.
5. Now rebase and what not to your hearts content.

It's a bit tedious, but it will get you to the desired outcome.

**Comment by kieranmaine** [comment]
Any plans to fix this github? Extremely frustrating.

**Comment by erkinalp** [comment]
Counter-proposal: Silently delete PR and retain commit comments after a force push without an ability to manually delete it.

**Comment by Spongman** [comment]
why force me to jump through all these hoops when i should just be able to click on the reopen button?

**Comment by FeepingCreature** [comment]
Force pushing the previous head commit, reopening, then force pushing the new commit worked fine for me. No gymnastics needed.

**Comment by virtualdxs** [comment]
This is ridiculous. I closed a PR, and I needed to retroactively sign my commit, so I rebased and force pushed, and then I ran into this.

**Comment by rob-miller** [comment]
I was able to get back to the state where the PR could be re-opened by force-pushing the the branch at the state when it was closed.
To get there I used `git reflog` and `git reflog --date=iso` to find the SHA1 of the last commit before the close, then the syntax
`git push origin <SHA1>:<branch_name_on_origin> --force`
where `origin` is my github repository's fork of the main repo where I closed the pull request.

**Comment by jxramos** [comment]
This might bet a good opportunity for a lil workflow tweak.

I hit this situation where I had a PR 2 based off the branch `A` which was under development in another PR 1. I was staging my work to roll into a parent branch that `A` would ultimately roll into and likewise for branch `B` in PR 2 when work on `A` wrapped up. I merged and deleted PR 1, and came to find that PR 2 was closed because it was targeting a deleted branch. Trying to reopen PR 2 yieled: _The A branch has been deleted_.

I initially targeted PR 2 off of A to better track the incremental ideas it captured without waiting for A to roll in first and let code review happen in parallel. It would serve us well if GitHub had a notion of "ultimate base" to fall back on whenever the active base is deleted. That would create a nice workflow, but maybe there's something flawed in my thinking that's shortsighted on something.

**Comment by lajarre** [comment]
@jxramos your development workflow makes total sense to me. I double the idea of a default-base, it would indeed come in handy in those cases.

In the meantime, you can rely on Github's REST API, and curl-up the pulls that are based on your branch before making the remote branch deletion. Along the lines of:

```
if [ $(curl https://api.github.com/repos/$repo/pulls?base=$branch | jq 'length > 0') = 'false' ];
then
    git push origin :$branch;
else
   echo "won't delete because $branch branch is a pulls base";
fi
```

**Comment by aspiers** [comment]
[**@Rob Miller**](https://github.com/rob-miller) commented on [December 3, 2019 1:06 PM](https://github.com/isaacs/github/issues/361#issuecomment-561158372 "2019-12-03T13:06:03Z - Replied by Github Reply Comments"):
> I was able to get back to the state where the PR could be re-opened by force-pushing the branch at the state when it was closed.
> To get there I used `git reflog` and `git reflog --date=iso` to find the SHA1 of the last commit before the close, then the syntax
> `git push origin <SHA1>:<branch_name_on_origin> --force`
> where `origin` is my github repository's fork of the main repo (I closed the pull request on the main repo).

If for some reason you lost the reflog, I found that you can still reconstruct the PR history via a GraphQL query on GitHub's v4 API, e.g.

```graphql
{
  repository(name: "straight.el", owner: "raxod502") {
    pullRequest(number: 564) {
      timelineItems(last: 10, itemTypes: HEAD_REF_FORCE_PUSHED_EVENT) {
        nodes {
          ... on HeadRefForcePushedEvent {
            afterCommit {
              url
            }
            createdAt
          }
        }
      }
    }
  }
}
```

If you feed that into docs.github.com/en/free-pro-team@latest/graphql/overview/explorer, you get:

```json
{
  "data": {
    "repository": {
      "pullRequest": {
        "timelineItems": {
          "nodes": [
            {
              "afterCommit": {
                "url": "https://github.com/raxod502/straight.el/commit/fae0871e878d3cbc34a93c5f3c7faa300093df1f"
              },
              "createdAt": "2020-08-30T14:12:49Z"
            },
            {
              "afterCommit": {
                "url": "https://github.com/raxod502/straight.el/commit/ee29a1e535fa78002b08d82f98c0369330339945"
              },
              "createdAt": "2020-08-30T14:50:25Z"
            }
          ]
        }
      }
    }
  }
}
```
